File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/gamePlay/DiceRollAction.java
package nl.hva.backend.domain.models.game.action.impl.gamePlay;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;

import javax.persistence.Entity;

/**
 * TODO: Add JavaDoc
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class DiceRollAction extends GameAction {
    private int firstDiceValue;
    private int secondDiceValue;

    public DiceRollAction(
            Player actor,
            int firstDiceValue,
            int secondDiceValue
    ) {
        super(actor);
        this.firstDiceValue = firstDiceValue;
        this.secondDiceValue = secondDiceValue;
    }

}

File: front-end/src/app/domain/models/game/action/impl/gamePlay/DiceRollAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class DiceRollAction extends GameAction {
  public constructor(
    id: number,
    actor: Player,
    performedAt: Date,
    firstDiceValue: number,
    secondDiceValue: number
  ) {
    super(DiceRollAction.name, id, actor, performedAt);
    this._firstDiceValue = firstDiceValue;
    this._secondDiceValue = secondDiceValue;
  }

  private _firstDiceValue: number;

  public get firstDiceValue(): number {
    return this._firstDiceValue;
  }

  private _secondDiceValue: number;

  public get secondDiceValue(): number {
    return this._secondDiceValue;
  }

  public static trueCopy(action: DiceRollAction): DiceRollAction | null {
    return new DiceRollAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.firstDiceValue,
      action.secondDiceValue
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "firstDiceValue": this.firstDiceValue,
        "secondDiceValue": this.secondDiceValue,
      });
  }

  public override toString(): string {
    return `${super.actor.user.username} rolled one ${this.firstDiceValue} and a ${this.secondDiceValue}`;
  }

}
